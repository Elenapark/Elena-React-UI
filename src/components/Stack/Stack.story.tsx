import React, { FC } from 'react';
import { Meta } from '@storybook/react';
import Stack, { CustomStackProps } from './Stack';

export default {
  title: 'Stack',
  component: Stack,
} as Meta<CustomStackProps>;

export const Default: FC = () => {
  return (
    <Stack backgroundColor="pink" spacing="24px">
      <span>Child1</span>
      <span>Child2</span>
      <span>Child3</span>
    </Stack>
  );
};
