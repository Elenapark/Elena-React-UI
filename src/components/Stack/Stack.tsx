import React, { ReactElement, ReactNode } from 'react';
import styled from 'styled-components';
import cx from 'classnames';

import { color, ColorProps, flexbox, FlexboxProps, layout, LayoutProps, space, SpaceProps } from 'styled-system';
import useRefinedChildren from '../../hooks/use-refined-children';

interface CustomStackProps {
  direction?: string;
  spacing?: number | string;
  className?: string;
  children?: ReactNode;
  [x: string]: any;
}

export type StackProps = CustomStackProps & SpaceProps & LayoutProps & ColorProps & FlexboxProps;

export const Stack: FCC<StackProps> = ({
  direction = 'column',
  spacing = '8px',
  className,
  children,
  ...props
}): ReactElement | null => {
  const refinedChildren = useRefinedChildren({ children, className });

  return (
    <StyledStack
      data-testid="stack-component"
      direction={direction}
      spacing={spacing}
      className={cx('stack', className)}
      {...props}
    >
      {refinedChildren}
    </StyledStack>
  );
};

const StyledStack = styled.div<StackProps>`
  ${space}
  ${layout} 
  ${color}  
  ${flexbox}
  display: flex;
  flex-direction: ${(props) => props.direction};
  gap: ${(props) => props.spacing};
`;
