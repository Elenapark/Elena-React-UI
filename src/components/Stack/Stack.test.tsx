import { render, screen } from '@testing-library/react';
import React from 'react';
import {Stack} from './Stack';

const invalidChildren = [1, 'text', <div>child1</div>];
const validChildren = [<div id="add">child1</div>, <div>child2</div>];

const renderStack = (children: any) => {
  const defaultProps = {
    direction: 'row',
    className: 'test',
    children,
  };
  render(<Stack {...defaultProps} />);
};

describe('<Stack />', () => {
  describe('when children valid', () => {
    it('should render', () => {
      renderStack(validChildren);
      expect(screen.getByTestId('stack-component')).toBeInTheDocument();
    });

    it('should render children', () => {
      renderStack(validChildren);
      expect(screen.getAllByTestId('stack-child-comp')).toHaveLength(2);
      expect(screen.getByText('child1')).toHaveProperty('id');
    });

    it('should allow custom classname', () => {
      renderStack(validChildren);
      expect(screen.getByTestId('stack-component')).toHaveProperty('className');
    });
  });

  describe('when children is invalid', () => {
    it('should make its children to be valid anyway', () => {
      renderStack(invalidChildren);
      expect(screen.queryAllByTestId('stack-child-comp')).toHaveLength(3);
    });
  });
});
