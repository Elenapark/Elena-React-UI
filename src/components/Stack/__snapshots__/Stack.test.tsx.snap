// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Stack /> when children valid should render 1`] = `
<div
  class="sc-bcXHqe eVfqAP stack test"
  data-testid="stack-component"
  direction="row"
  spacing="8px"
>
  <div
    data-testid="stack-child-comp"
    id="add"
  >
    child1
  </div>
  <div
    data-testid="stack-child-comp"
  >
    child2
  </div>
</div>
`;
